File: examples/legacy/cs/gmsub_attn/gmsub_attn.cs
/*
 * Copyright 2007-2014 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * All Rights Reserved.
 */

/*! \file gmsub_attn.cs
    This file contains a simple example subscriber using the GMSEC .NET API.

*/

using System;
using System.Linq;
using System.IO;
using System.Diagnostics;
using System.Collections.Generic;
using System.Configuration;
using System.Management;

using System.Net.Sockets;

namespace gmsub_attn
{
    using Nasa.Gmsec.Net;
    using Gmsec.Example.Common;

    class gmsub_attn : GmsecExample
    {
        private readonly string ATTENTION_DELIMITER_DEFAULT = "***";

        private Config config;
        private Connection conn;
        private Message message;
        private String delimiterString;
        private String logLevel = "INFO";
        private List<string> subjects;
        private Queue<string> includeFields;

        public gmsub_attn()
        {
        }

        private void printUsage()
        {
            Console.Out.WriteLine(
                "\nusage: gmsub_attn_cs connectiontype=<middleware> " +
                "[ <parameter>=<value> ]\n" +
                "\n\tNote that the parameter 'connectiontype' is required. " +
                "\n\tThe rest of other parameters are optional.\n" +
                "\n\tserver=<server name> " +
                "(required if middleware is not bolt/MB locally)" +
                "\n\tsubject=<subject name>" +
                "\n\tsubject.<N>=<subject name>" +
                "\n\t\tNote: N must be 1 or greater, and be consecutive." +
                "\n\t\tFor example: subject.1=GMSEC.A subject.2=GMSEC.B ..." +
                "\n\titerations=<iterations>" +
                "\n\tmsg_timeout_ms=<timeout period (milliseconds)>" +
                "\n\tprog_timeout_s=<timeout period (seconds)>" +
                "\n\t\tNote: msg_timeout_ms must be defined to use prog_timeout_s" +
                "\n\tloglevel=<log level>" +
                "\n\tcfgfile=<config_filename.xml>" +
				"\n\tattn-port=<Attention! server port>" +
				"\n\tattn-server=<Attention! server host>" +
				"\n\tinclude-fields=<comma-separated list of field names," +
				"\n\t\tthe content of which will be sent to the Attention!" +
				"\n\t\tserver>" +
				"\n\tdelimiter-string=<Attention! server string delimiter>" +
				"\n\nFor more information, see API User's Guide\n\n");
        }

        private void cleanUp()
        {
            if (conn.IsConnected())
            {
                checkNoThrow("Disconnect", conn.Disconnect());
            }
            checkNoThrow("Destroying Connection", ConnectionFactory.Destroy(conn));
        }

        private void Run(string[] args)
        {
            config = new Config(args);

            addToConfigFromFile(config);

            string strServer, strPort;

            if (isOptionInValid(config, args, true))
            {
                printUsage();
                return;
            }

            Status status = new Status();

            // Output GMSEC API version
            Console.Out.WriteLine(ConnectionFactory.GetAPIVersion());

            // Create the connection
            status = ConnectionFactory.Create(config, out conn);
            check("Error creating connection", status);

            // Connect
            status = conn.Connect();
            check("Error connecting", status);

            // Determine the log level
            config.GetValue("LOGLEVEL", out logLevel);

            //Determine the subjects to listen to
            subjects = determineSubjects(config);

            // Determine the message field names to insert into the delimited string
            // in the case of an overflow situation (>1000 characters)
            includeFields = determineIncludeFields(config);

            delimiterString = determineDelimiterString(config);

            //Determine how many messages to recieve
            int iterations = get(config, "ITERATIONS", 0);

            if (iterations > 0)
            {
                Console.Out.WriteLine("Waiting for up to " + iterations + " messages");
            }

            //Determine Timeout Periods
            int msg_timeout_ms = get(config, "MSG_TIMEOUT_MS", EX_MSG_TIMEOUT_MS);
            int prog_timeout_s = get(config, "PROG_TIMEOUT_S", EX_PROG_TIMEOUT_S);

            Console.Out.WriteLine("Middleware version = " + conn.GetLibraryVersion());

            //Subscribe
            for (int i = 0; i < subjects.Count; i++)
            {
                Console.Out.WriteLine("Subscribing to " + subjects[i]);
                status = conn.Subscribe(subjects[i]);
                check("Subscribe", status);
            }

            bool done = false;
            int count = 0;
            int intPort = 0;

            Stopwatch stopWatch = new Stopwatch();

            try
            {
                strPort = get(config, "ATTN-PORT", "4999");
                intPort = Convert.ToInt32(strPort);
            }
            catch (FormatException fe)
            {
                Console.WriteLine("FormatException : {0}", fe.ToString());
                Console.WriteLine("Value input for ATTN-PORT is not a valid sequence of digits.");
                Environment.Exit(1);
            }
            catch (OverflowException oe)
            {
                Console.WriteLine("OverflowException : {0}", oe.ToString());
                Console.WriteLine("The number cannot fit in an Int32.");
                Environment.Exit(1);
            }

            strServer = get(config, "ATTN-SERVER", "127.0.0.1");

            int connectionAttempts = 0;

            while (connectionAttempts < 3)
            {
                System.Net.IPHostEntry ipHostInfo = System.Net.Dns.GetHostEntry(strServer);
                System.Net.IPAddress ipAddress = ipHostInfo.AddressList[0];
                System.Net.IPEndPoint ipEndPoint = new System.Net.IPEndPoint(ipAddress, intPort);

                Socket socket = new Socket(ipAddress.AddressFamily, SocketType.Stream, ProtocolType.Tcp);

                try
                {
                    socket.Connect(ipEndPoint);
                }
                catch (ArgumentNullException ae)
                {
                    Console.WriteLine("ArgumentNullException : {0}", ae.ToString());
                    Environment.Exit(1);
                }
                catch (SocketException se)
                {
                    Console.WriteLine("SocketException : {0}", se.ToString());
                    Console.WriteLine("Trying to connect again (Attempt " + ++connectionAttempts + " of 3)", se.ToString());
                    System.Threading.Thread.Sleep(500);
                    //Environment.Exit(1);
                    continue;
                }
                catch (Exception e)
                {
                    Console.WriteLine("Unexpected exception : {0}", e.ToString());
                    Environment.Exit(1);
                }

                connectionAttempts = 0;

                //strServer = ConfigurationManager.AppSettings["ATTNSRV"];
                Console.Out.WriteLine("Connected to Server: " + strServer + ":" + intPort);

                while (!done)
                {
                    if (prog_timeout_s != (int)GMSECWaitDefs.WAIT_FOREVER)
                    {
                        if (!stopWatch.IsRunning)
                        {
                            stopWatch.Start();
                        }
                        else if (stopWatch.ElapsedMilliseconds >= (long)(prog_timeout_s * 1000))
                        {
                            Console.Out.WriteLine("Program Timeout Reached!");
                            stopWatch.Stop();
                            break;
                        }
                    }

                    status = conn.GetNextMsg(out message, msg_timeout_ms);

                    if (status.IsError())
                    {
                        //do not exit on timeouts
                        if (status.GetCode() == GMSECErrorCodes.TIMEOUT_OCCURRED)
                        {
                            Console.Out.WriteLine("Timeout Occurred");
                            continue;
                        }
                        else
                        {
                            check("GetNextMsg", status);
                        }
                    }

                    if (message.IsValid())
                    {
                        count++;

                        if (iterations > 0 && count >= iterations)
                        {
                            done = true;
                        }

                        string xml;
                        string subject;
                        string attnArgs;
						
						// Get the subject of the message
                        status = message.GetSubject(out subject);
                        check("GetSubject", status);

                        // Convert the GMSEC message to an XML String
                        status = message.ToXML(out xml);
                        check("ToXML", status);

                        if (xml.Length != 0)
                        {
                            if (logLevel.Equals("DEBUG", StringComparison.CurrentCultureIgnoreCase)
							&& (subject.Equals("GMSEC.GPM.GPM1.MSG.LOG.WDOG", StringComparison.CurrentCultureIgnoreCase)
							|| subject.Equals("GMSEC.GPM.GPM1.MSG.C2CX.LSIMSS.HB", StringComparison.CurrentCultureIgnoreCase)))
                            {
                            	Console.Out.WriteLine("Received Message:\n" + xml);
                            }

                            // Create the delimited string that will be sent to the Attention Generic Agent
                            attnArgs = getAttnArgString(message);

                            /*
                            if (logLevel.Equals("DEBUG", StringComparison.CurrentCultureIgnoreCase))
                            {
                                Console.Out.WriteLine("Sending the string to the Attention! Server:\n" + attnArgs);
                            }
                            */

                            byte[] data = System.Text.Encoding.UTF8.GetBytes(attnArgs);
                            try
                            {
                                int bytesSent = 0;

                                while (bytesSent < data.Length)
                                {
                                    // Blocks until send returns.
                                    int result = socket.Send(data, bytesSent, data.Length - bytesSent, SocketFlags.None);

                                    if (result > 0)
                                    {
                                        bytesSent += result;
                                    }
                                    else if (result < 0)
                                    {
                                        Console.Out.WriteLine("Tried to write more data than was in the buffer!");
                                    }
                                }

                                // Wait for an ack (Operation blocks until data is received, no need for other logic)
                                int bytesRec = socket.Receive(data);
                            }
                            catch (SocketException e)
                            {
                                Console.WriteLine("{0} Error code: {1}.", e.Message, e.ErrorCode);
                                break;
                            }

                            /*
                            //strPort = My.settings.ATTNPRT;
                            Environment.SetEnvironmentVariable("ATTNSRV", strServer, EnvironmentVariableTarget.Process);
                            Environment.SetEnvironmentVariable("ATTNPRT", strPort, EnvironmentVariableTarget.Process);
                            Process attn = new Process();
                            attn.StartInfo.FileName = "attn.exe";
                            attn.StartInfo.Arguments = attnArgs;
                            attn.StartInfo.WindowStyle = System.Diagnostics.ProcessWindowStyle.Hidden;
                            attn.Start();
                            */
                        }
						else
						{
							Console.WriteLine("Unable to get the XML-format of a message with the subject: {0}" + subject);
						}

                        if (subject.Length != 0)
                        {
                            if (subject.Equals("GMSEC.TERMINATE"))
                            {
                                done = true;
                            }
                        }

                        status = conn.DestroyMessage(message);
                        check("Destroy Message", status);
                    }
					else
					{
						Console.WriteLine("Retrieved an invalid message from the GMSEC Bus");
					}
                }
            }

            cleanUp();
        }

        private string getAttnArgString(Message message)
        {
            Field tmpField;
            string tmpFieldValue;
            // 5.0;gpm-auto1|<First field>|3***<Rest of the message fields>
            string attnArgs = "5.0;gpm-auto1|";
            int counter = 0;
            bool firstTime = true;

            foreach (string fieldName in includeFields)
            {

                if (fieldName == "SUBJECT")
                {
                    message.GetSubject(out tmpFieldValue);
                }
                else
                {
                    GMSECTypeDefs fType;
                    message.GetField(fieldName, out tmpField);
                    tmpField.GetType(out fType);

                    switch (fType)
                    {
                        case GMSECTypeDefs.BOOL:
                            bool boolValue;
                            tmpField.GetValue(out boolValue);
                            tmpFieldValue = boolValue.ToString();
                            break;
                        case GMSECTypeDefs.CHAR:
                            sbyte charValue;
                            tmpField.GetValue(out charValue);
                            tmpFieldValue = charValue.ToString();
                            break;
                        case GMSECTypeDefs.F32:
                            float floatValue;
                            tmpField.GetValue(out floatValue);
                            tmpFieldValue = floatValue.ToString();
                            break;
                        case GMSECTypeDefs.F64:
                            double doubleValue;
                            tmpField.GetValue(out doubleValue);
                            tmpFieldValue = doubleValue.ToString();
                            break;
                        case GMSECTypeDefs.I16:
                            short shortValue;
                            tmpField.GetValue(out shortValue);
                            tmpFieldValue = shortValue.ToString();
                            break;
                        case GMSECTypeDefs.I32:
                            int intValue;
                            tmpField.GetValue(out intValue);
                            tmpFieldValue = intValue.ToString();
                            break;
                        case GMSECTypeDefs.I64:
                            long longValue;
                            tmpField.GetValue(out longValue);
                            tmpFieldValue = longValue.ToString();
                            break;
                        case GMSECTypeDefs.STRING:
                            tmpField.GetValue(out tmpFieldValue);
                            break;
                        case GMSECTypeDefs.BIN:
                        case GMSECTypeDefs.COMPOUND:
                        case GMSECTypeDefs.I8:
                        case GMSECTypeDefs.U16:
                        case GMSECTypeDefs.U32:
                        case GMSECTypeDefs.U64:
                        case GMSECTypeDefs.U8:
                        case GMSECTypeDefs.UNSET:
                            tmpFieldValue = "NO " + fieldName;
                            break;
                        // TODO: Unsigned short, int, long, unset, bin, compound
                    }
                    tmpField.GetValue(out tmpFieldValue);
                }

                //if (tmpFieldValue != null || tmpFieldValue.Length != 0)
                if (!String.IsNullOrEmpty(tmpFieldValue))
                {
                    attnArgs += tmpFieldValue;
                }
                else
                {
                    attnArgs += "NO " + fieldName;
                }

                if (firstTime)
                {
                    attnArgs += "|3";
                    firstTime = false;
                }

                if (++counter < includeFields.Count)
                {
                    attnArgs += delimiterString;
                }
            }

            return attnArgs;
        }

        // Parse out the overflow field names from the comma-delimited string
        public Queue<string> determineIncludeFields(Config cfg)
        {
            Queue<string> fields = new Queue<string>();
            string inclFields = get(cfg, "INCLUDE-FIELDS");
            string[] inclFieldsSplit = inclFields.Split(',');

            Console.Out.WriteLine("\nThe following fields will be concatenated:");
            foreach (string fieldName in inclFieldsSplit)
            {
                if (fieldName.Length != 0)
                {
                    Console.Out.WriteLine(fieldName);
                    fields.Enqueue(fieldName);
                }
            }
            Console.Out.WriteLine();

            return fields;
        }

        // Determine the delimiter used in the formatted string
        public string determineDelimiterString(Config cfg)
        {
            string delimiterString = get(cfg, "DELIMITER-STRING");

            if (delimiterString == null)
            {
                delimiterString = ATTENTION_DELIMITER_DEFAULT;
            }

            Console.Out.WriteLine("Delimiter string is set to: " + delimiterString + "\n");
            return delimiterString;
        }

        static void Main(string[] args)
        {
            //if (Properties.Settings.Default.CONSOLE)
            //{
            //}
            //else
            //{
            //    TextWriterTraceListener myWriter = new
            //    TextWriterTraceListener(System.Console.Out);
            //    Debug.Listeners.Add(myWriter);
            //    Debug.WriteLine("Test output 1 ");
            //    Stream myFile = File.Create("output.txt");
            //    TextWriterTraceListener myTextListener = new
            //    TextWriterTraceListener(myFile);
            //    Debug.Listeners.Add(myTextListener);
            //    Debug.WriteLine("Test output 2 ");
            //    Debug.WriteLine("Test output 3 ");
            //    myWriter.Flush();
            //    myWriter.Close();
            //    myFile.Flush();
            //    myFile.Close();
            //}

            try
            {
                gmsub_attn exProg_gmsub_attn = new gmsub_attn();
                exProg_gmsub_attn.Run(args);
            }

            catch (Exception e)
            {
                Console.Out.WriteLine(e.ToString());
            }

            Console.Out.WriteLine("Press <enter> to exit");
            Console.Read();
        }
    }
}

File: examples/legacy/java/gmsub_attn.java
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */






/** 
 * @file gmsub_attn.java
 *
 * An example Java GMSEC subscriber which packetizes data to be sent
 *   to an Attention! generic collector agent.
 */

import gov.nasa.gsfc.gmsecapi.Config;
import gov.nasa.gsfc.gmsecapi.Connection;
import gov.nasa.gsfc.gmsecapi.ConnectionFactory;
import gov.nasa.gsfc.gmsecapi.Field;
import gov.nasa.gsfc.gmsecapi.Message;
import gov.nasa.gsfc.gmsecapi.Status;
import gov.nasa.gsfc.gmsecapi.GMSEC_Bin;
import gov.nasa.gsfc.gmsecapi.GMSEC_Bool;
import gov.nasa.gsfc.gmsecapi.GMSEC_Char;
import gov.nasa.gsfc.gmsecapi.GMSEC_Double;
import gov.nasa.gsfc.gmsecapi.GMSEC_F32;
import gov.nasa.gsfc.gmsecapi.GMSEC_F64;
import gov.nasa.gsfc.gmsecapi.GMSEC_Float;
import gov.nasa.gsfc.gmsecapi.GMSEC_I8;
import gov.nasa.gsfc.gmsecapi.GMSEC_I16;
import gov.nasa.gsfc.gmsecapi.GMSEC_I32;
import gov.nasa.gsfc.gmsecapi.GMSEC_I64;
import gov.nasa.gsfc.gmsecapi.GMSEC_Long;
import gov.nasa.gsfc.gmsecapi.GMSEC_Short;
import gov.nasa.gsfc.gmsecapi.GMSEC_String;
import gov.nasa.gsfc.gmsecapi.GMSEC_U16;
import gov.nasa.gsfc.gmsecapi.GMSEC_U32;
import gov.nasa.gsfc.gmsecapi.GMSEC_ULong;
import gov.nasa.gsfc.gmsecapi.GMSEC_UShort;
import static gov.nasa.gsfc.gmsecapi.gmsecAPI.*;
import gov.nasa.gsfc.gmsecapi.util.*;

import java.util.List;
import java.util.ArrayList;

import java.io.IOException;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.nio.charset.Charset;
import java.net.Socket;
import java.net.UnknownHostException;

public class gmsub_attn implements Example
{
	private final String ATTENTION_DELIMITER_DEFAULT = "***";

	StatusChecker checker;
	Config config;
	Connection connection;
	Message message;
	String delimiterString = "";
	List<String> includeFields;

	gmsub_attn(Config config) throws ExampleException {
		checker = new StatusChecker();
		this.config = config;
		Util.initialize(config);

		if (!Util.hasConnectionType(config))
		{
			printUsage();
			throw new ExampleException("Invalid option");
		}
	}

	private void printUsage() {
		Log.Error("\nusage: java gmsub_attn connectionType=<middleware> " +
				"[ <parameter>=<value> ]\n" +
				"\n\tNote that the parameters 'connectiontype' is required. "  +
				"\n\tThe rest of other parameters are optional.\n" +
				"\n\tserver=<server name> " + 
				"(required if middleware is not bolt/MB locally)" +
				"\n\tsubject=<subject name>" +
				"\n\titerations=<iterations>" +
				"\n\tsubject.<N>=<sub-subject name>" +
				"\n\t\tNote: N must be 1 or greater" +
				"\n\t\tFor example, subject.1=GMSEC.A subject.2=GMSEC.B" +
				"\n\tmsg_timeout_ms=<timeout period (milliseconds)>" +
				"\n\tprog_timeout_s=<timeout period (seconds)>" +
				"\n\t\tNote: msg_timeout_ms must be specified for prog_timeout_s" +
				"\n\tloglevel=<log level>" +
				"\n\tcfgfile=<config_filepath.xml>" +
				"\n\tattn-port=<Attention! server port>" +
				"\n\tattn-server=<Attention! server host>" +
				"\n\tinclude-fields=<comma-separated list of field names," +
				"\n\t\tthe content of which will be sent to the Attention!" +
				"\n\t\tserver>" +
				"\n\tdelimiter-string=<Attention! server string delimiter>" +
				"\n\nFor more information, see API User's Guide\n\n");
	}

	public boolean run() throws ExampleException
	{
		Status status;
		String strHost, strPort;

		// determine what subjects to listen to
		List<String> subjects = new ArrayList<String>(10);

		String value = Util.get(config,"SUBJECT");
		if (value != null)
		{
			subjects.add(value);
		}
		else
		{
			for (int i = 1; ; ++i)
			{
				String tmp = Util.get(config, "SUBJECT." + i);
				if (tmp != null)
				{
					subjects.add(tmp);
				}
				else
				{
					break;
				}
			}
		}

		if (subjects.isEmpty()) {
			String subject = Util.get(config, "SUBJECT", "GMSEC.>");
			subjects.add(subject);
		}
		else {
			subjects.add("GMSEC.TERMINATE");
		}
		
		// Determine the message field names to insert into the delimited string
		// in the case of an overflow situation (>1000 characters)
		includeFields = determineIncludeFields(config);

		delimiterString = determineDelimiterString(config);

		int iterations = Util.get(config, "ITERATIONS", 0);
		if (iterations > 0)
		{
			Log.Info("waiting for up to " + iterations + " messages");
		}

		int msg_timeout_ms = Util.get(config, "MSG_TIMEOUT_MS", Util.EX_MSG_TIMEOUT_MS);
		int prog_timeout_s = Util.get(config, "PROG_TIMEOUT_S", Util.EX_PROG_TIMEOUT_S);
        if(msg_timeout_ms == GMSEC_WAIT_FOREVER && prog_timeout_s != GMSEC_WAIT_FOREVER)
        {
           Log.Error("Usage: To use prog_timeout_s, msg_timeout_ms must be defined!");
           return false;
        }                          

		// Establish a connection
		connection = Util.openConnection(config);

		// Output GMSEC API version
		Log.Info(ConnectionFactory.GetAPIVersion());
		
		// Output middleware version
		Log.Info("Middleware version = " + connection.GetLibraryVersion());

		// Subscribe
		for (String subject : subjects) {
			Log.Info("subscribing to " + subject);
			checker.calling("Subscribe(" + subject + ")").check(connection.Subscribe(subject));
		}

		// Wait for messages
		int count = 0;
		boolean done = false;
		long prevTime = System.currentTimeMillis();
		long nextTime = 0;
		int elapsedTime = 0;	//ms
		
		int intPort = 0;

		try
		{
			strPort = Util.get(config, "ATTN-PORT", "4999");
			intPort = Integer.parseInt(strPort);
		}
		catch (NumberFormatException nfe)
		{
			Log.Error("Encountered a NumberFormatException: " + nfe.toString());
			Log.Error("Value input for ATTN-PORT is not a valid sequence of digits.");
			System.exit(1);
		}

		strHost = Util.get(config, "ATTN-SERVER", "127.0.0.1");
		
		int connectionAttempts = 0;

		while (connectionAttempts < 3)
		{
			connectionAttempts++;
			Socket socket;

			try
			{
				Log.Info("Connecting to server: " + strHost + ":" + intPort);
				socket = new Socket(strHost, intPort);

				connectionAttempts = 0;
				
				while (!done) {
				
					if(elapsedTime >= (prog_timeout_s * 1000)
										 && prog_timeout_s != GMSEC_WAIT_FOREVER)
					{
						done = true;
						Log.Info("Program Timeout Reached");
						continue;
					}
				
					message = new Message();
					
					status = connection.GetNextMsg(message, msg_timeout_ms);
					
					if (status.IsError()) {
						if (status.GetCode() == GMSEC_TIMEOUT_OCCURRED)
						{
							Log.Info("timeout occurred");
						}
						else
						{
							checker.checkNoThrow(status);
						}
					}
					
					// Get the subject of the message
					GMSEC_String gmsecSubject = new GMSEC_String();
					checker.calling("GetSubject").checkNoThrow(message.GetSubject(gmsecSubject));
					String subject = gmsecSubject.Get();
					
					nextTime = System.currentTimeMillis();
					elapsedTime += (int) (nextTime - prevTime);
					prevTime = nextTime;
					
					if (message.IsValid())
					{
						count++;
						
						if (iterations > 0 && count >= iterations)
						{
							done = true;
						}

						// Convert the GMSEC message to an XML String
						GMSEC_String gmsecXml = new GMSEC_String();
						status = message.ToXML(gmsecXml);
						String xml = gmsecXml.Get();

						if (!status.IsError())
						{
							Log.Debug("gmsub_attn Received Message:\n" + xml);

							// Create the delimited string that will be sent to the Attention Generic Agent
							String attnArgs = getAttnArgString(message, subject);
							
							byte[] data = attnArgs.getBytes("UTF-8");
							
							try
							{
								DataOutputStream out = new DataOutputStream(socket.getOutputStream());
								DataInputStream in = new DataInputStream(socket.getInputStream());
								
								// Send the byte-encrypted data to the server
								out.write(data, 0, data.length);

								// Wait for an ack (Operation blocks until data is received, no need for other logic)
								in.read(data);
								
								data = null;
							}
							catch (IOException ioe)
							{
								Log.Error("Encountered an IOException: " + ioe.toString());
								break;
							}
						}
						else
						{
							Log.Error("gmsub_attn Unable to retrieve the XML string from the message using Message::ToXML()");
						}

						Message tmp = message;
						message = null;
						checker.calling("DestroyMessage").checkNoThrow(connection.DestroyMessage(tmp));
					}
					else
					{
						Log.Error("gmsub_attn Received an invalid message according to a function call to Message::IsValid()");
					}
				}
			}
			catch (UnknownHostException uhe)
			{
				Log.Error("Encountered an UnknownHostException: " + uhe.toString());
				Log.Error("Unable to connect to the Attention! server, trying again.");
				continue;
			}
			catch (IOException ioe)
			{
				Log.Error("Encountered an IOException: " + ioe.toString());
				System.exit(1);
			}
		}

		return true;
	}

	
	private String getAttnArgString(Message message, String subject) throws ExampleException
	{
		// 5.0;gpm-auto1|<First field>|3***<Rest of the message fields>
		String attnArgs = "5.0;gpm-auto1|";
		int counter = 0;
		boolean firstTime = true;

		for (String fieldName : includeFields)
		{
			String tmpFieldValue = "";
			
			if (fieldName.equalsIgnoreCase("SUBJECT"))
			{
				tmpFieldValue = subject;
			}
			else
			{
				Field tmpField = new Field();
				message.GetField(fieldName, tmpField);
				
				if (tmpField != null)
				{
					GMSEC_UShort gmsecType = new GMSEC_UShort();
					tmpField.GetType(gmsecType);
					int fType = gmsecType.Get();

					switch (fType)
					{
						case GMSEC_TYPE_BOOL:
							GMSEC_Bool booleanValue = new GMSEC_Bool();
							tmpField.GetValueBOOL(booleanValue);
							tmpFieldValue = booleanValue.toString();
							break;
						case GMSEC_TYPE_CHAR:
							GMSEC_Char charValue = new GMSEC_Char();
							tmpField.GetValueCHAR(charValue);
							tmpFieldValue = charValue.toString();
							break;
						case GMSEC_TYPE_F32:
							GMSEC_F32 floatValue = new GMSEC_F32();
							tmpField.GetValueF32(floatValue);
							tmpFieldValue = floatValue.toString();
							break;
						case GMSEC_TYPE_F64:
							GMSEC_F64 doubleValue = new GMSEC_F64();
							tmpField.GetValueF64(doubleValue);
							tmpFieldValue = doubleValue.toString();
							break;
						case GMSEC_TYPE_I16:
							GMSEC_I16 shortValue = new GMSEC_I16();
							tmpField.GetValueI16(shortValue);
							tmpFieldValue = shortValue.toString();
							break;
						case GMSEC_TYPE_I32:
							GMSEC_I32 intValue = new GMSEC_I32();
							tmpField.GetValueI32(intValue);
							tmpFieldValue = intValue.toString();
							break;
						case GMSEC_TYPE_I64:
							GMSEC_I64 longValue = new GMSEC_I64();
							tmpField.GetValueI64(longValue);
							tmpFieldValue = longValue.toString();
							break;
						case GMSEC_TYPE_STRING:
							GMSEC_String stringValue = new GMSEC_String();
							tmpField.GetValueSTRING(stringValue);
							tmpFieldValue = stringValue.Get();
							break;
						case GMSEC_TYPE_BIN:
						case GMSEC_TYPE_I8:
						case GMSEC_TYPE_U16:
						case GMSEC_TYPE_U32:
							tmpFieldValue = "NO " + fieldName;
							break;
						// TODO: Unsigned short, int, long, unset, bin, compound
					}
				}
				else
				{
					tmpFieldValue = "NO " + fieldName;
				}
			}

			if (tmpFieldValue != null && tmpFieldValue.length() != 0)
			{
				attnArgs += tmpFieldValue;
			}
			else
			{
				attnArgs += "NO " + fieldName;
			}

			if (firstTime)
			{
				attnArgs += "|3";
				firstTime = false;
			}

			if (++counter < includeFields.size())
			{
				attnArgs += delimiterString;
			}
		}

		return attnArgs;
	}
	
	
	// Parse out the overflow field names from the comma-delimited string
	public ArrayList<String> determineIncludeFields(Config cfg)
	{
		ArrayList<String> fields = new ArrayList<String>();
		String inclFields = Util.get(cfg, "INCLUDE-FIELDS");
		if (inclFields != null && inclFields.length() > 0)
		{
			String[] inclFieldsSplit = inclFields.split(",");

			Log.Info("\nThe following fields will be concatenated:");
			for (String fieldName : inclFieldsSplit)
			{
				if (fieldName.length() != 0)
				{
					Log.Info(fieldName);
					fields.add(fieldName);
				}
			}
			Log.Info("");
		}

		return fields;
	}

	// Determine the delimiter used in the formatted string
	public String determineDelimiterString(Config cfg)
	{
		String delimiterString = Util.get(cfg, "DELIMITER-STRING");

		if (delimiterString == null)
		{
			delimiterString = ATTENTION_DELIMITER_DEFAULT;
		}

		Log.Info("Delimiter string is set to: " + delimiterString + "\n");
		return delimiterString;
	}


	public boolean cleanup() throws ExampleException {
		if (message != null && message.isValid()) {
			Status status = connection.DestroyMessage(message);
			Util.checkNoThrow("DestroyMessage", status);
		}

		if (connection != null)
			Util.closeConnection(connection);

		return true;
	}

	public static void main(String[] args) {

		Example example = null;

		try
		{
			Config config = new Config(args);

			Util.addToConfigFromFile(config);

			example = new gmsub_attn(config);
			example.run();
		}
		catch (ExampleException e)
		{
			e.printStackTrace();
		}
		finally
		{
			if (example != null)
			{
				Util.cleanup(example);
			}
		}

	}

}

